SYCL/ESIMD/api/simd_copy_to_from.cpp

```diff
@@ -0,0 +1,223 @@
+             Flags) {
+  std::cout << "  Running accessor " << Title << " test, N=" << N << "...\n";
+
+  for (int I = 0; I < N; ++I) {
+    Src[I + Off] = I + 1;
+    Dst[I + Off] = 0;
+  }
+
+  try {
+    buffer<T, 1> SrcB(Src, range<1>(Off + N));
+    buffer<T, 1> DstB(Dst, range<1>(Off + N));
+
+    Q.submit([&](handler &CGH) {
+       auto SrcA = SrcB.template get_access<access::mode::read>(CGH);
+       auto DstA = DstB.template get_access<access::mode::write>(CGH);
+
+       CGH.parallel_for(sycl::range<1>{1}, [=](id<1>) SYCL_ESIMD_KERNEL {
+         simd<T, N> Vals;
+         Vals.copy_from(SrcA, Off * sizeof(T), Flags{});
+         Vals.copy_to(DstA, Off * sizeof(T), Flags{});
+       });
+     }).wait();
+  } catch (cl::sycl::exception const &E) {
+    std::cout << "ERROR. SYCL exception caught: " << E.what() << std::endl;
+    return false;
+  }
+
+  unsigned NumErrs = 0;
+  for (int I = 0; I < N; ++I)
+    if (Dst[I + Off] != Src[I + Off])
+      if (++NumErrs <= 10)
+        std::cout << "failed at " << I << ": " << Dst[I + Off]
+                  << " (Dst) != " << Src[I + Off] << " (Src)\n";
+
+  std::cout << (NumErrs == 0 ? "    Passed\n" : "    FAILED\n");
+  return NumErrs == 0;
+}
+
+template <typename T, int N> bool testUSM(const std::string &Type, queue &Q) {
+  struct Deleter {
+    queue Q;
+    void operator()(T *Ptr) {
+      if (Ptr) {
+        sycl::free(Ptr, Q);
+      }
+    }
+  };
+
+  std::unique_ptr<T, Deleter> Src(sycl::aligned_alloc_shared<T>(1024u, 512u, Q),
+                                  Deleter{Q});
+  std::unique_ptr<T, Deleter> Dst(sycl::aligned_alloc_shared<T>(1024u, 512u, Q),
+                                  Deleter{Q});
+
+  constexpr unsigned VecAlignOffset = esimd::detail::getNextPowerOf2<N>();
+
+  bool Pass = true;
+
+  Pass &= testUSM<T, N>(Q, Src.get(), Dst.get(), VecAlignOffset + 1u,
+                        Type + " element_aligned", element_aligned);
+  Pass &= testUSM<T, N>(Q, Src.get(), Dst.get(), VecAlignOffset,
+                        Type + " vector_aligned", vector_aligned);
+  Pass &= testUSM<T, N>(Q, Src.get(), Dst.get(), 128u / sizeof(T),
+                        Type + " overaligned<128>", overaligned<128u>);
+
+  return Pass;
+}
+
+template <typename T> bool testUSM(const std::string &Type, queue &Q) {
+  bool Pass = true;
+
+  Pass &= testUSM<T, 1>(Type, Q);
+  Pass &= testUSM<T, 2>(Type, Q);
+  Pass &= testUSM<T, 3>(Type, Q);
+  Pass &= testUSM<T, 4>(Type, Q);
+
+  Pass &= testUSM<T, 7>(Type, Q);
+  Pass &= testUSM<T, 8>(Type, Q);
+
+  Pass &= testUSM<T, 15>(Type, Q);
+  Pass &= testUSM<T, 16>(Type, Q);
+
+  if constexpr (sizeof(T) < 8) {
+    Pass &= testUSM<T, 24>(Type, Q);
+    Pass &= testUSM<T, 25>(Type, Q);
+
+    Pass &= testUSM<T, 31>(Type, Q);
+    Pass &= testUSM<T, 32>(Type, Q);
+  }
+
+  return Pass;
+}
+
+template <typename T, int N> bool testAcc(const std::string &Type, queue &Q) {
+  std::unique_ptr<T> Src(new T[512u]);

```

Comment:
strictly speaking, alignment of `new` result is not known, so either aligned allocation should be used or buffer start should be enforced to known alignment ((ptr + align) & ~align)

Reply:
For SVM test for sure. But is it required for the accessor as well?

Reply:
accessor is wrapped around the pointer allocated by `new` in the test, so I think yes, it is.
For buffer w/o host pointer it is likely that allocated memory would be page-aligned, but I'm not sure if this is guaranteed or not, so even in that case forcing alignment would make sense (esp. overaligned).

Reply:
Okay, it that is the case I will change this to aligned allocation.